executor/src/witgen/jit/affine_symbolic_expression.rs

```diff
@@ -0,0 +1,550 @@
+    /// Otherwise, we assert that they are different.
+    pub expected_equal: bool,
+}
+
+impl<T: FieldElement, V> Assertion<T, V> {
+    pub fn assert_is_zero(condition: SymbolicExpression<T, V>) -> Effect<T, V> {
+        Self::assert_eq(condition, SymbolicExpression::from(T::from(0)))
+    }
+    pub fn assert_is_nonzero(condition: SymbolicExpression<T, V>) -> Effect<T, V> {
+        Self::assert_neq(condition, SymbolicExpression::from(T::from(0)))
+    }
+    pub fn assert_eq(lhs: SymbolicExpression<T, V>, rhs: SymbolicExpression<T, V>) -> Effect<T, V> {
+        Effect::Assertion(Assertion {
+            lhs,
+            rhs,
+            expected_equal: true,
+        })
+    }
+    pub fn assert_neq(
+        lhs: SymbolicExpression<T, V>,
+        rhs: SymbolicExpression<T, V>,
+    ) -> Effect<T, V> {
+        Effect::Assertion(Assertion {
+            lhs,
+            rhs,
+            expected_equal: false,
+        })
+    }
+}
+
+/// Represents an expression `a_1 * x_1 + ... + a_k * x_k + offset`,
+/// where the `a_i` and `offset` are symbolic expressions, i.e. values known at run-time,
+/// and the `x_i` are unknown variables.
+#[derive(Debug, Clone)]
+pub struct AffineSymbolicExpression<T: FieldElement, V> {
+    coefficients: BTreeMap<V, SymbolicExpression<T, V>>,
+    offset: SymbolicExpression<T, V>,
+}
+
+impl<T: FieldElement, V: Display> Display for AffineSymbolicExpression<T, V> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if self.coefficients.is_empty() {
+            write!(f, "{}", self.offset)
+        } else {
+            write!(
+                f,
+                "{}",
+                self.coefficients
+                    .iter()
+                    .map(|(var, coeff)| if coeff.is_known_one() {
+                        var.to_string()
+                    } else if coeff.is_known_minus_one() {
+                        format!("-{var}")
+                    } else {
+                        format!("{coeff} * {var}")
+                    })
+                    .join(" + ")
+            )?;
+            if !self.offset.is_known_zero() {
+                write!(f, " + {}", self.offset)?;
+            }
+            Ok(())
+        }
+    }
+}
+
+impl<T: FieldElement, V> From<SymbolicExpression<T, V>> for AffineSymbolicExpression<T, V> {
+    fn from(k: SymbolicExpression<T, V>) -> Self {
+        AffineSymbolicExpression {
+            coefficients: Default::default(),
+            offset: k,
+        }
+    }
+}
+
+impl<T: FieldElement, V: Ord + Clone + Display> AffineSymbolicExpression<T, V> {
+    pub fn from_known_variable(var: V) -> Self {
+        SymbolicExpression::from_var(var).into()
+    }
+    pub fn from_unknown_variable(var: V) -> Self {
+        AffineSymbolicExpression {
+            coefficients: [(var, T::from(1).into())].into_iter().collect(),
+            offset: SymbolicExpression::from(T::from(0)),
+        }
+    }
```

Comment:
I wondered about that as well, but in the end, both will be `Cell`s, i.e. references to a cell in the trace. Some of them are known, some of them are not, and it will even change while we progress.

In the end, what I call `variable` here is just another name for a "terminal" in the symbolic system that is not a constant known value.